RFC-0022-Model-Partitioning-in-Pipeline-Parallelism.md

```diff
@@ -0,0 +1,255 @@
+
+Our approach here starts with an approach similar to fairscale, which only do top-level partition, but it could be improved further by exploring more partition approaches, i.e. with operation level partitioning by torch.fx, we could partition the model in a more even manner, which further improves the efficiency. Since we only expose a simple API and return the partitioned model with stages assigned to the devices, the internal implementation can be improved as we explore more ideas. The plan for partition algorithm exploration:
+
+* Use fairscale/gpipe's [Block Partition of Sequences](https://arxiv.org/pdf/1308.2452.pdf) as our first step
+* See if we can apply [Fiduccia-Mattheyses Heuristic](https://en.wikipedia.org/wiki/Fiduccia%E2%80%93Mattheyses_algorithm) to partition the graph and how the performance compare with the default one
+* See if we can do operation-level tracing, and partition the model into several balanced `torch.fx.GraphModule` instead of using the original module architecture
+
+### Approach that enables more possibility of balanced partitioning
+
+* fx trace the graph, collect the statistics of each node
+* partition the graph using a bottom up approach: starting from treating each node as a partition, and merge them if it lowers the cost, use BFS to scan the merge order
+* The result will be an `fx.GraphModule` for each partition, the original module architecture is not preserved
+    * qual names of param might change
+
+Partition:
+Dict[fx.graphmodule, remotedevice]
+
+in order to apply the algorithm, we need to assign each top-level submodule with a proper cost, we calculate the cost based on memory and execution costs.
+
+
+### Cost calculation
+
+The cost of a submodule is defined as the following:
+
+```
+cost(m) = time_weight * execution_time(m) 
+            + memory_weight * memory_cost(m)
+
+memory_cost(m) = parameter cost * optimizer_multiplier 
+                    + activation_cost
+```
+
+where `memory_weight + time_weight = 1`. How to weigh between time and memory cost? Undecided, a heuristic number `time_weight=0.6`
+
+* @mrshenli: regarding balancing time between memory, I kind of feel we should prioritize time over memory. Because people usually use Pipeline parallel to accelerate training, and the training speed the ultimate goal. Balanced execution time has a directly impact on total pipeline makespan (suppose one phase is slower than others, then all other devices will need to wait for that phase). If the above assumption is correct, it looks like we should first try to balance time, and only try to balance memory when a shard of the optimally time-balanced model cannot fit in some devices. 
```

Comment:
People use pipeline parallelism often to train large models like GPT-3, which don't fit on a single accelerator. I think of memory_cost as a constraint rather than an objective (minimize time-per-iteration while ensuring that any worker's memory footprint does not exceed its memory capacity).